sway-lsp/src/core/token.rs

```diff
@@ -101,81 +116,340 @@ pub fn traverse_node(node: AstNode, tokens: &mut Vec<Token>) {
+            ..
         } => {
-            let ident = name.suffix;
-            let token = Token::from_ident(&ident, TokenType::FunctionApplication);
+            handle_expression(*condition, tokens);
+            handle_expression(*then, tokens);
+            if let Some(r#else) = r#else {
+                handle_expression(*r#else, tokens);
+            }
+        }
+        Expression::MatchExp { if_exp, .. } => {
+            handle_expression(*if_exp, tokens);
+        }
+        Expression::AsmExpression { .. } => {
+            //TODO handle asm expressions
+        }
+        Expression::MethodApplication {
+            method_name,
+            arguments,
+            contract_call_params,
+            ..
+        } => {
+            let ident = method_name.easy_name();
+            let token = Token::from_ident(&ident, TokenType::MethodApplication);
             tokens.push(token);
 
             for exp in arguments {
                 handle_expression(exp, tokens);
             }
+
+            //TODO handle methods from imported modules
+            if let MethodName::FromType {
+                type_name: Some(type_name),
+                ..
+            } = &method_name
+            {
+                handle_custom_type(type_name, tokens);
+            }
+
+            for field in contract_call_params {
+                let token = Token::from_ident(
+                    &field.name,
+                    TokenType::StructExpressionField(get_struct_field_details(&ident)),
+                );
+                tokens.push(token);
+                handle_expression(field.value, tokens);
+            }
+        }
+        Expression::SubfieldExpression { prefix, .. } => {
+            handle_expression(*prefix, tokens);
+            //TODO handle field_to_access?
+        }
+        Expression::DelineatedPath {
+            call_path, args, ..
+        } => {
+            for prefix in call_path.prefixes {
+                //TODO find the correct token type for this!
+                let token = Token::from_ident(&prefix, TokenType::DelineatedPath);
+                tokens.push(token);
+            }
+
+            let token = Token::from_ident(&call_path.suffix, TokenType::DelineatedPath);
+            tokens.push(token);
+
+            for exp in args {
+                handle_expression(exp, tokens);
+            }
+        }
+        Expression::AbiCast {
+            abi_name, address, ..
+        } => {
+            let ident = abi_name.suffix;
+            let token = Token::from_ident(&ident, TokenType::AbiCast);
+            tokens.push(token);
+
+            handle_expression(*address, tokens);
+        }
+        Expression::ArrayIndex { prefix, index, .. } => {
+            handle_expression(*prefix, tokens);
+            handle_expression(*index, tokens);
+        }
+        Expression::DelayedMatchTypeResolution { .. } => {
+            //Should we handle this since it gets removed during type checking anyway?
```

Comment:
Good Q: Is this how the parser represents the `_` syntax? I.e. `let foo: Vec<_> = ...` where `_` explicitly indicates that the type should be inferred? If so, it would be cool if `sway-lsp` could show what type is inferred when you hover over it. That said, that would all have to wait til we introduce type checking anyways.

Reply:
Yeah this would be cool, might need to wait for type checking as you say.